config-example.yml

```diff
@@ -159,18 +159,23 @@ graphhopper:
 
 
 
-  ##### Spatial Rules #####
-  # Spatial Rules require some configuration and only work with the DataFlagEncoder.
-
-
-  # Spatial Rules require you to provide Polygons in which the rules are enforced
-  # The line below contains the default location for the files which define these borders
-  # spatial_rules.borders_directory: core/files/spatialrules
-
-  # You can define the maximum BBox for which spatial rules are loaded.
-  # You might want to do this if you are only importing a small area and don't need rules for other countries.
-  # Having less rules, might result in a smaller graph. The line below contains the world-wide bounding box, uncomment and adapt to your need.
-  # spatial_rules.max_bbox: -180,180,-90,90
+  ##### Custom Areas & Spatial Rules #####
+
+  # Custom Areas are loaded from user provided GeoJSON files. A registered SpatialRule is enforced for an area if it has the same ID.
+  # The following fields can be configured for each custom area file:
+  # - location (required): the path of the GeoJSON file
+  # - id_field (required): they GeoJSON property of the (multi)polygons to be used as the ID of the resulting area
+  # - encoded_value (optional, default: ""): the encoded value to be used if the area should be stored in the graph
+  # - encoded_value_limit (optional, default: -1): expected number of encountered areas to be stored in the encoded value. Defaults to the number of areas if -1.
+  # - max_bbox (optional, default: -180,180,-90,90): maximum BBox for which custom areas are loaded. Having less areas, might result in a smaller graph.
+  custom_area_files: []
+  
+  # custom_area_files:
+  #   - location: core/files/spatialrules/countries.geojson
+  #     id_field: "ISO3166-1:alpha3"
+  #     encoded_value: country
+  #     encoded_value_limit: 63
+  #     max_bbox: -180,180,-90,90
```

Comment:
Is it possible to move the encoded value definition to `graph.encoded_values`? Then we could define it like `encoded_values=country|limit=63,surface,...` and we have all explicit encoded values in one place and would need to use it for the custom area.

And is max_bbox still necessary? I think I saw some PR where this requirement was removed, but maybe I mix something up.

Reply:
> Is it possible to move the encoded value definition to `graph.encoded_values`? Then we could define it like `encoded_values=country|limit=63,surface,...` and we have all explicit encoded values in one place and would need to use it for the custom area.

I wasn't sure which option is better. From a configuration standpoint it was nice to have it in one place. But i think you're right and it's easier to configure/maintain if we manage it with the other EVs.

> And is max_bbox still necessary? I think I saw some PR where this requirement was removed, but maybe I mix something up.

It's an optimization which allows to skip country boundaries which are not relevant for the imported dataset. e.g importing an europe.pbf while using the global country boundaries file in our repo.

Reply:
@karussell should we keep some kind of prefix for all custom area EVs?

e.g `encoded_values=custom_area.country|limit=63,custom_area.state|limit=127,surface,...`

My current approach is to create EncodedValues with a namespace using `EncodingManger.getKey(String,String)`. This makes it easier to retrieve the correct EV for the PathDetailsBuilder later:

```java
for (String checkPrefix : requestedPathDetails) {
     if (checkPrefix.startsWith(CustomArea.key("")) && evl.hasEncodedValue(checkPrefix))
          builders.add(new StringDetails(checkPrefix, evl.getStringEncodedValue(checkPrefix)));
}
```
Adding a prefix looks a bit verbose but without a it we'd have to treat any unknown key as a custom area encoded value and that would also include typos...


If we want to add other types of "dynamic" EVs in the future, we have to come up with a manageable solution.
We *could* default to creating StringEncodedValues for unknown keys but

Reply:
> It's an optimization which allows to skip country boundaries which are not relevant for the imported dataset. e.g importing an europe.pbf while using the global country boundaries file in our repo.

Can you explain what you mean here? Do you mean geometries outside of max_bbox will be skipped?

> should we keep some kind of prefix for all custom area EVs?

Yes, problem we need some identification like we have for RouteNetwork but a dot is not possible. I suggest a postfix like there e.g. `country_area`? But then it would be ugly in our script `in_area_modena_area` :D ... so probably we should change it there too? (probably even independent of this PR to merge it fast)

Reply:
> in_area_country_area

Ah, no. This would not be a problem. Still I would change it to `in_xy_area` for consistency I guess.

Reply:
> > It's an optimization which allows to skip country boundaries which are not relevant for the imported dataset. e.g importing an europe.pbf while using the global country boundaries file in our repo.
> 
> Can you explain what you mean here? Do you mean geometries outside of max_bbox will be skipped?

Exactly. Editing large geojson files is rather error prone so the user might for example use a big geojson file with lots of features while he only needs a specific subset for the small pbf dataset he wants to import. e.g using our world boundaries file while importing a DACHL pbf

I've just added a commit to also take this into account if the user doesn't provide a `encoded_value_limit`. The size of the resulting encoded value is therefore determined by the number of custom areas intersecting the bounding box.

Note that the bbox is also used to modify the polygons by calculating an intersection:

https://github.com/graphhopper/graphhopper/blob/4f126833b0d962b68e2dc103e748903ec08951ff/core/src/main/java/com/graphhopper/routing/util/area/CustomAreaHelper.java#L110-L123

> 
> > should we keep some kind of prefix for all custom area EVs?
> 
> Yes, probably we need some identification like we have for RouteNetwork but a dot is not possible. I suggest a postfix like there e.g. `country_area`? But then it would be ugly in our script `in_area_country_area` :D ... so probably we should change it there too? (probably even independent of this PR to merge it fast)

`_area` looks good